# Introduction to Python for Bioinformatics and the Life Sciences

## Final Seminar Project Report 

### Author: **[Ajay Sah](https://github.com/aajusah98/)**
### Matriculation Number : 23172531

## 2. Data Source

### Datasource1: Normalized bulk gene expression data from breast cancer patients here
- Data URL: [Link to dataset](https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-BRCA.htseq_fpkm.tsv.gz)
- Data Type: TSV

### Datasource2: Corresponding phenotype data here:
- Data URL: [Link to dataset](https://gdc-hub.s3.us-east-1.amazonaws.com/download/TCGA-BRCA.GDC_phenotype.tsv.gz)
- Data Type: TSV

##### Installing Required All Librarie

In [1]:
# Install pandas
%pip install pandas

# Install numpy
%pip install numpy

# Install seaborn
%pip install seaborn

# Install matplotlib
%pip install matplotlib


Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


##### Importing All Required Librarie In This Project 

In [2]:
import pandas as pd

## Step 1: Load And Filter Data
 
 1.1 **Load both datasets into pandas DataFrame objects.**
 
   - Use sample IDs as indices and gene IDs as columns. If the data is not provided in this format, transpose it if necessary.

In [3]:
phenotype_df= pd.read_csv('TCGA-BRCA.GDC_phenotype.tsv', sep="\t",index_col=0)
gene_expression_df=pd.read_csv('TCGA-BRCA.htseq_fpkm.tsv',sep="\t")

- Printing all the columns in  the phenotype data to know the features of the data which is given by `phenotype_df.columns`

- Printing the shape of the phenotype data to know the shape of the data which is given by `phenotype_df.shape`

- Printing the index of the phenotype data to know the indices which is given by `phenotype_df.index` 

- `phenotype_df.head(10)` Print the first 10 data from the data frame 

In [4]:
print(phenotype_df.columns)
print(phenotype_df.shape)
print(phenotype_df.index)
print(phenotype_df.head(10))

Index(['additional_pharmaceutical_therapy', 'additional_radiation_therapy',
       'additional_surgery_locoregional_procedure',
       'additional_surgery_metastatic_procedure',
       'age_at_initial_pathologic_diagnosis',
       'axillary_lymph_node_stage_method_type',
       'axillary_lymph_node_stage_other_method_descriptive_text',
       'batch_number', 'bcr', 'bcr_followup_barcode',
       ...
       'days_to_collection.samples', 'days_to_sample_procurement.samples',
       'initial_weight.samples', 'is_ffpe.samples', 'oct_embedded.samples',
       'preservation_method.samples', 'sample_type.samples',
       'sample_type_id.samples', 'state.samples', 'tissue_type.samples'],
      dtype='object', length=139)
(1284, 139)
Index(['TCGA-A2-A0CY-01A', 'TCGA-B6-A40B-01A', 'TCGA-AO-A0J8-01A',
       'TCGA-A8-A08J-01A', 'TCGA-E2-A14N-01A', 'TCGA-BH-A1F0-01A',
       'TCGA-BH-A1F0-11B', 'TCGA-BH-A0BZ-01A', 'TCGA-BH-A0BZ-11A',
       'TCGA-AC-A2FM-01A',
       ...
       'TCGA-A8-A07L-01A',

- Printing all the columns in  the gene expression  data to know the features of the data which is given by `phenotype_df.columns`

- Printing the shape of the gene expression  data to know the shape of the data which is given by `phenotype_df.shape`

- Printing the index of the gene expression data to know the indices which is given by `phenotype_df.index` 

- `phenotype_df.head(10)` Print the first 10 data from the data frame 

In [5]:
print(gene_expression_df.columns)
print(gene_expression_df.shape)
print(gene_expression_df.index)
print(gene_expression_df.head(10))

Index(['Ensembl_ID', 'TCGA-E9-A1NI-01A', 'TCGA-A1-A0SP-01A',
       'TCGA-BH-A1EU-11A', 'TCGA-A8-A06X-01A', 'TCGA-E2-A14T-01A',
       'TCGA-AC-A8OS-01A', 'TCGA-A8-A09K-01A', 'TCGA-OL-A5RY-01A',
       'TCGA-BH-A0DG-01A',
       ...
       'TCGA-BH-A0DT-11A', 'TCGA-E9-A1R0-01A', 'TCGA-BH-A0B6-01A',
       'TCGA-B6-A0RN-01A', 'TCGA-A8-A09W-01A', 'TCGA-EW-A1P3-01A',
       'TCGA-A7-A13F-11A', 'TCGA-A2-A0T6-01A', 'TCGA-A7-A5ZW-01A',
       'TCGA-BH-A203-01A'],
      dtype='object', length=1218)
(60483, 1218)
RangeIndex(start=0, stop=60483, step=1)
           Ensembl_ID  TCGA-E9-A1NI-01A  TCGA-A1-A0SP-01A  TCGA-BH-A1EU-11A  \
0   ENSG00000242268.2          0.091708          0.000000          0.057899   
1   ENSG00000270112.3          0.019573          0.004701          0.016302   
2  ENSG00000167578.15          2.235898          1.863334          1.704753   
3   ENSG00000273842.1          0.000000          0.000000          0.000000   
4   ENSG00000078237.5          2.321945          4.226

- As we notice theat index of the gene expression df is numerical RangeIndex(start=0, stop=60483, step=1)  but we need want sample id as indices and gene ID

- we can simply tranpose the dataframe let see what we will get after transposing

- we use `gene_expression_df.T.reset_index()` for transposing it 

In [6]:

# Transpose the Gene Expression  DataFrame
gene_expression_df = gene_expression_df.T.reset_index()


In [7]:
print(gene_expression_df.head(10))
print(gene_expression_df.index)

              index                  0                  1                   2  \
0        Ensembl_ID  ENSG00000242268.2  ENSG00000270112.3  ENSG00000167578.15   
1  TCGA-E9-A1NI-01A           0.091708           0.019573            2.235898   
2  TCGA-A1-A0SP-01A                0.0           0.004701            1.863334   
3  TCGA-BH-A1EU-11A           0.057899           0.016302            1.704753   
4  TCGA-A8-A06X-01A                0.0                0.0            1.947481   
5  TCGA-E2-A14T-01A                0.0                0.0             2.73469   
6  TCGA-AC-A8OS-01A           0.354334                0.0            2.397119   
7  TCGA-A8-A09K-01A                0.0           0.005787            2.337327   
8  TCGA-OL-A5RY-01A                0.0                0.0            2.256976   
9  TCGA-BH-A0DG-01A                0.0                0.0            1.968791   

                   3                  4                   5  \
0  ENSG00000273842.1  ENSG00000078237.5  ENSG

- After transposing we can see that row and colums are swaped but still we have a problem which is numeric vlaues on index and colmun

- .iloc[0]: This selects the first row of the DataFrame, which you are using to set as column headers.

- .drop(0): This function removes the first row from the DataFrame. 

In [8]:
# Set the first row as column headers
gene_expression_df.columns=gene_expression_df.iloc[0] 

# Drop the first row from the DataFrame
gene_expression_df=gene_expression_df.drop(0)

# set the index
gene_expression_df.set_index('Ensembl_ID', inplace=True)

- Printng index,columns, shape and dataframe head to verify if we are on  good track 
- and our dataframe look good now , now sample id is as indices and gene ids for gene expression 

In [9]:
print(gene_expression_df.columns)
print(gene_expression_df.shape)
print(gene_expression_df.index)
print(gene_expression_df.head(10))

Index(['ENSG00000242268.2', 'ENSG00000270112.3', 'ENSG00000167578.15',
       'ENSG00000273842.1', 'ENSG00000078237.5', 'ENSG00000146083.10',
       'ENSG00000225275.4', 'ENSG00000158486.12', 'ENSG00000198242.12',
       'ENSG00000259883.1',
       ...
       'ENSG00000238244.3', 'ENSG00000186115.11', 'ENSG00000216352.1',
       'ENSG00000267117.1', 'ENSG00000273233.1', 'ENSG00000105063.17',
       'ENSG00000231119.2', 'ENSG00000280861.1', 'ENSG00000123685.7',
       'ENSG00000181518.3'],
      dtype='object', name=0, length=60483)
(1217, 60483)
Index(['TCGA-E9-A1NI-01A', 'TCGA-A1-A0SP-01A', 'TCGA-BH-A1EU-11A',
       'TCGA-A8-A06X-01A', 'TCGA-E2-A14T-01A', 'TCGA-AC-A8OS-01A',
       'TCGA-A8-A09K-01A', 'TCGA-OL-A5RY-01A', 'TCGA-BH-A0DG-01A',
       'TCGA-B6-A0I9-01A',
       ...
       'TCGA-BH-A0DT-11A', 'TCGA-E9-A1R0-01A', 'TCGA-BH-A0B6-01A',
       'TCGA-B6-A0RN-01A', 'TCGA-A8-A09W-01A', 'TCGA-EW-A1P3-01A',
       'TCGA-A7-A13F-11A', 'TCGA-A2-A0T6-01A', 'TCGA-A7-A5ZW-01A',
       '

## Step 1: Load And Filter Data
 
 1.2 **Partition the gene expression data frame into two data frames.**
 
   - such that one data
frame only contains data for primary tumor samples and the other one only contains data
for healthy tissue samples. You find the information necessary to split the data in the
phenotype data. To identify the variable which contains the necessary information on the
sample types, you can have a look at this paper: https://doi.org/10.1093/bib/bbad413
(especially the Methods section).

- As we have to divide our phonedata in two type healthy tissues sample, primary tissues sample we will use `sample_type.samples` feature which conntain the type of the tissues,

- print`phenotype_df['sample_type.samples'].unique()` it will give all the unique values at sample type column

- As we can see that it have 3 differnet type, 

    i. Solid Tissue Normal :- It describes tissue samples that are collected from healthy, non-cancerous regions of the body
    
    ii. Primary Tumor :- Primary tumors are the initial site of cancer growth and are crucial in the diagnosis and treatment planning of cancer

    iii. Metastatic :- Metastatic cancer is classified and treated based on the type of the primary cancer
    
    
- There is nan, let drop that nan data

In [10]:
print(phenotype_df['sample_type.samples'].unique())

['Primary Tumor' 'Solid Tissue Normal' 'Metastatic' nan]


In [11]:
#Dropping nan values from pheno_data
phenotype_df.dropna(subset=['sample_type.samples'], inplace=True)
sample_types = phenotype_df['sample_type.samples'].unique()
print(sample_types)

['Primary Tumor' 'Solid Tissue Normal' 'Metastatic']


-  We will filter all the row which which  contain `Primary Tumor` from the phonetype_df and we will store there

- Use `intersection` function to find the matchining indices in both phonetype_df and gene_expression_df and 
`difference` for notfound indieces 

- Use `found_indices_for_primary_tumor_samples` variable to extract all the data from gene expression 
- .loc[] is used here for label-based indexing. It will access rows from gene_expression_df whose indices match those in found_indices_for_primary_tumor_samples.

In [12]:
# Filter gene expression data based on phenotype data
primary_tumor_samples = phenotype_df[phenotype_df['sample_type.samples'] == 'Primary Tumor']



found_indices_for_primary_tumor_samples=primary_tumor_samples.index.intersection(gene_expression_df.index)

not_found_indices_for_primary_tumor_samples = primary_tumor_samples.index.difference(gene_expression_df.index)

# filter the data according to found indicies at primary_tumor_samples
primary_tumor_expression_data = gene_expression_df.loc[found_indices_for_primary_tumor_samples]
# print(gene_expression.index)

print(primary_tumor_expression_data)


0                ENSG00000242268.2 ENSG00000270112.3 ENSG00000167578.15  \
TCGA-A2-A0CY-01A               0.0          0.022873           1.561889   
TCGA-B6-A40B-01A               0.0               0.0           2.452261   
TCGA-AO-A0J8-01A               0.0          0.003354           2.100417   
TCGA-A8-A08J-01A               0.0               0.0           1.731999   
TCGA-E2-A14N-01A               0.0               0.0           1.306598   
...                            ...               ...                ...   
TCGA-A1-A0SF-01A               0.0               0.0           2.067434   
TCGA-PE-A5DD-01A               0.0               0.0            3.31852   
TCGA-E2-A10B-01A               0.0          0.007769           2.036247   
TCGA-BH-A0E0-01A               0.0               0.0           1.863133   
TCGA-BH-A0B2-01A               0.0               0.0           1.854281   

0                ENSG00000273842.1 ENSG00000078237.5 ENSG00000146083.10  \
TCGA-A2-A0CY-01A        

In [14]:
# Filter gene expression data based on phenotype data
healthy_tissue_samples = phenotype_df[phenotype_df['sample_type.samples'] == 'Solid Tissue Normal']


found_indices_for_healthy_tissue_samples=primary_tumor_samples.index.intersection(gene_expression_df.index)

not_found_indices_for_found_indices_for_healthy_tissue_samples = healthy_tissue_samples.index.difference(gene_expression_df.index)

# filter the expression data according to found indicies at healty tissue samples
healthy_tissue_expression_data = gene_expression_df.loc[found_indices_for_healthy_tissue_samples]

# print(gene_expression.index)

print(healthy_tissue_expression_data)


0                ENSG00000242268.2 ENSG00000270112.3 ENSG00000167578.15  \
TCGA-A2-A0CY-01A               0.0          0.022873           1.561889   
TCGA-B6-A40B-01A               0.0               0.0           2.452261   
TCGA-AO-A0J8-01A               0.0          0.003354           2.100417   
TCGA-A8-A08J-01A               0.0               0.0           1.731999   
TCGA-E2-A14N-01A               0.0               0.0           1.306598   
...                            ...               ...                ...   
TCGA-A1-A0SF-01A               0.0               0.0           2.067434   
TCGA-PE-A5DD-01A               0.0               0.0            3.31852   
TCGA-E2-A10B-01A               0.0          0.007769           2.036247   
TCGA-BH-A0E0-01A               0.0               0.0           1.863133   
TCGA-BH-A0B2-01A               0.0               0.0           1.854281   

0                ENSG00000273842.1 ENSG00000078237.5 ENSG00000146083.10  \
TCGA-A2-A0CY-01A        